### PR TITLE
Decouple `CrateNum` of `CrateInfo` from the compilation session

### PR DESCRIPTION
**Pull request by 0dvictor**
`CrateNum` used inside a `CrateInfo` is self-contained and hence can be decoupled from the compilation session. By doing so, a compilation session can consume `CrateInfo` generated from a different compilation session. Ultimately, it allows splitting linker invocation into a separate compilation session.

### CODE REVIEW

File: src/librustc_codegen_ssa/lib.rs [hunk 1]
```diff
@@ -117,6 +116,19 @@ bitflags::bitflags! {
     }
 }
 
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct CrateNum(pub u32);
```
Comment (bjorn3) [inline src/librustc_codegen_ssa/lib.rs]: This name can be confused with `def_id::CrateNum`.

File: src/librustc_codegen_ssa/base.rs [hunk 2]
```diff
@@ -708,6 +709,17 @@ fn assert_and_save_dep_graph(tcx: TyCtxt<'_>) {
...
+impl From<def_id::CrateNum> for CrateNum {
+    fn from(crate_num: def_id::CrateNum) -> Self {
+        CrateNum(crate_num.as_u32())
+    }
+}
+impl Into<def_id::CrateNum> for CrateNum {
+    fn into(self) -> def_id::CrateNum {
+        def_id::CrateNum::from_u32(self.0)
+    }
+}
```
Comment (bjorn3) [inline src/librustc_codegen_ssa/base.rs]: Please move these to `lib.rs` below the definition of `crate::CrateNum`.

File: src/librustc_codegen_ssa/lib.rs [hunk 3]
```diff
@@ -117,6 +116,19 @@ bitflags::bitflags! {
     }
 }
 
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
```
Comment (bjorn3) [inline src/librustc_codegen_ssa/lib.rs]:
```suggestion
#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, RustcDecodable, RustcEncodable)]
```

**Comment by bjorn3** [comment]
Not a member of `rust-lang`, so your `r?` didn't work.

**Review by bjorn3 — APPROVED** [review]
This is indeed what I had in mind in https://github.com/rust-lang/rust/pull/67195#discussion_r360645260.

File: src/librustc_codegen_ssa/lib.rs (hunk 3 above)
Comment (0dvictor) [reply src/librustc_codegen_ssa/lib.rs]: IMO adding `RustcDecodable` and `RustcEncodable` is beyond the scope of this PR. I prefer to leave it to #67195, where `RustcDecodable` and `RustcEncodable` are added to some other structs and enums.

File: src/librustc_codegen_ssa/base.rs (hunk 2 above)
Comment (0dvictor) [reply src/librustc_codegen_ssa/base.rs]: Done.

**Comment by bjorn3** [comment]
@0dvictor Could you please add changes as extra commits instead of force-pushing them while this PR is still being reviewed. Quoting `@rust-highfive`:

> If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code.

File: src/librustc_codegen_ssa/lib.rs (hunk 1 above)
Comment (0dvictor) [reply src/librustc_codegen_ssa/lib.rs]: Good point. I chose to keep the name `CrateNum` mainly because I feel `LinkingCrateNum` is misleading. `LinkingCrateNum` gives an impression that this "crate num" is solely for the purpose of linking, but in fact it is just for `CrateInfo` to index multiple crates. In addition, this new `rustc_codegen_ssa::CrateNum` serves the same purpose as `def_id::CrateNum` does under different scope; so I decided to keep its name `CrateNum`. While reviewing this PR, we can think about it more and find it a better name.

**Comment by 0dvictor** [comment]
> Could you please add changes as extra commits instead of force-pushing them while this PR is still being reviewed. Quoting `@rust-highfive`:
>> If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code.

Apologize for my mistake and the inconvenience. I'll definitely add any further changes as extra commits.

**Comment by 0dvictor** [comment]
Rebased to resolve merge conflicts.

**Comment by 0dvictor** [comment]
Rebased to resolve merge conflicts.

**Comment by 0dvictor** [comment]
Rebased again to fix the error introduced by the previous rebase.

File: src/librustc_codegen_ssa/lib.rs [hunk 4]
```diff
@@ -118,6 +118,29 @@ bitflags::bitflags! {
...
+impl From<def_id::CrateNum> for CrateNum {
+    fn from(crate_num: def_id::CrateNum) -> Self {
+        CrateNum(crate_num.as_u32())
+    }
+}
+impl Into<def_id::CrateNum> for CrateNum {
+    fn into(self) -> def_id::CrateNum {
+        def_id::CrateNum::from_u32(self.0)
+    }
+}
```
Comment (bjorn3) [inline src/librustc_codegen_ssa/lib.rs]: Is this actually used?

Comment (0dvictor) [reply src/librustc_codegen_ssa/lib.rs]: Yes, it is used in src/librustc_codegen_ssa/back/write.rs:995.
`each_linked_rlib_for_lto.push((cnum.into(), path.to_path_buf()));`

Comment (bjorn3) [reply src/librustc_codegen_ssa/lib.rs]: I think you should use this `CrateNum` for `each_linked_rlib_for_lto` and `exported_symbols` too.

Comment (0dvictor) [reply src/librustc_codegen_ssa/lib.rs]: Good point, updated.

**Comment by 0dvictor** [comment]
Rebased to resolve merge conflicts.

**Comment by 0dvictor** [comment]
Rebased again to fix the error introduced by the previous rebase.

File: src/librustc_codegen_ssa/base.rs [hunk 5]
```diff
@@ -835,9 +839,6 @@ impl CrateInfo {
             if tcx.is_profiler_runtime(cnum) {
                 info.profiler_runtime = Some(cnum.into());
             }
-            if tcx.is_sanitizer_runtime(cnum) {
-                info.sanitizer_runtime = Some(cnum.into());
-            }
```
Comment (Mark-Simulacrum) [inline src/librustc_codegen_ssa/base.rs]: Why was this removed?

**Review by Mark-Simulacrum — COMMENTED** [review]
It is unclear to me what benefits this gives us. `CrateNum` in rustc_hir is already fairly decoupled -- I would like to see some documentation as to the difference between this CrateNum and the one in rustc_hir. We seem to be just converting into the newly introduced CrateNum in this PR, which seems odd, as then I would expect it to be the same... but if it is, then why is there a different type?

**Comment by Mark-Simulacrum** [comment]
cc @Centril

I'm going to assign myself as reviewer here though depending on the outcome of my questions may reassign to someone else (or suggest e.g. compiler team feedback via a design meeting).

**Comment by 0dvictor** [comment]
> It is unclear to me what benefits this gives us. CrateNum in rustc_hir is already fairly decoupled -- I would like to see some documentation as to the difference between this CrateNum and the one in rustc_hir. We seem to be just converting into the newly introduced CrateNum in this PR, which seems odd, as then I would expect it to be the same... but if it is, then why is there a different type?

This PR is to resolve [this comment](https://github.com/rust-lang/rust/pull/67195#discussion_r360642681) of PR #67195. I think @bjorn3's concern is `def_id::CrateNum` is coupled with the compilation session, so that it is not conceptually correct when serializing `CodegenResults` then de-serializing it into a different compilation session. To resolve the issue, this PR creates a local `CrateNum` within the `CodegenResults` . The proposed `CrateNum` is no longer tied with a compilation session. and hence moving `CodegenResults` across compilation sessions is conceptually correct.

PR #67195 is the first step to allow splitting linker invocation from the compilation process. Therefore, one invocation of `rustc` serializes `CodegenResults` to a file in the compilation pass; then another invocation of `rustc` would de-serialize the file back to a `CodegenResults`. The two invocations would be two different compilation session.

@bjorn3: could you double check if I understand your original concern correctly?

File: src/librustc_codegen_ssa/base.rs (hunk 5 above)
Comment (0dvictor) [reply src/librustc_codegen_ssa/base.rs]: Sorry about the confusion, these three lines was incorrectly kept due to rebase error. They were deleted in PR #65241.

**Comment by bjorn3** [comment]
> I think @bjorn3's concern is def_id::CrateNum is coupled with the compilation session, so that it is not conceptually correct when serializing CodegenResults then de-serializing it into a different compilation session.

Indeed, there should not be an instance of `libserialize::Deserialize` which doesn't remap the `CrateNum` as required by the current `Session`. Because `Session` will likely not exist when the linking is performed in a separate rustc invocation, this means that `CrateNum` as is can't be used for serializing the information required by the linking step.

File: src/librustc_codegen_ssa/base.rs (hunk 5 above)
Comment (0dvictor) [reply src/librustc_codegen_ssa/base.rs]: I've rebased this PR again, so these lines no longer shows up as "deleted".

**Comment by Mark-Simulacrum** [comment]
Reading the discussion on that PR (https://github.com/rust-lang/rust/pull/67195#discussion_r360642681) is somewhat helpful, in concert with your comment.

However, I still remain skeptical that this is the right approach.

> In this PR, crate numbers are used within one CodegenResults, and CodegenResults only belongs to one specific session; therefore, crate numbers are unique. Then we can use CodegenResults.crate_info.used_crate_source to find the crate from a crate number.

This seems incomplete -- what are you doing with the CrateNums in that crate_info data structure? I would personally expect that if we're serializing to something like `*.rlink` file we'd want to encode the CrateInfo struct in a form that doesn't have crate nums (e.g., only the file paths that need to be linked or something like that).

**Comment by 0dvictor** [comment]
> what are you doing with the CrateNums in that crate_info data structure?

It serves as unique identifiers for the crate's dependencies. Such unique identifiers do not have to be `CrateNum` - anything that can differentiate the dependencies would work (e.g. SVH, file path, counters etc.), but nonetheless `CrateNum` is the most handy one.

`crate_info.used_crate_source` provides a mapping table from `CrateNum` to `CrateSource` containing all dependent crates. Other fields of `crate_info` map `CrateNum` to the dependent crate's properties, for example, `crate_info.compiler_builtins` indicate one of the crates if a compiler builtin. Therefore, information inside `crate_info` is self-contained and can be used in different compilation sessions.

> I would personally expect that if we're serializing to something like *.rlink file we'd want to encode the CrateInfo struct in a form that doesn't have crate nums (e.g., only the file paths that need to be linked or something like that).

Totally agree, but we still need some identifier for each dependent. Such identifier is effectively equivalent to `CrateNum`. For example, we could encode the `.rlink` file as following format:
```
Section used_crate_source
/path/to/dylib0 | /path/to/rlib0 | /path/to/rmeta0
/path/to/dylib1 | /path/to/rlib1 | /path/to/rmeta1
...
/path/to/dylibX | /path/to/rlibX | /path/to/rmetaX
-------------------------------------------------------
Section panic_runtime
None
-------------------------------------------------------
Section compiler_builtins
used_crate_source.entry0
-------------------------------------------------------
Section native_libraries
used_crate_source.entry5 | libc, libm
used_crate_source.entry9 | libc, pthread
-------------------------------------------------------
... other sections
-------------------------------------------------------
```
Entry indices to `used_crate_source` are effectively equivalent to `CrateNum`.

In addition, when decoding the `.rlink` file, we still have to somehow create a `CrateNum` unless we are open to heavily refactor existing linking code so that it doesn't use `CrateNums`.

**Comment by Mark-Simulacrum** [comment]
Instead of having a separate CrateNum (i.e. this PR) can we do the mapping inside the serialization step for CrateInfo? AFAICT, we can deserialize and serialize CrateNum (into just u32, if necesssary) already -- just as well as this new CrateNum -- that feels sufficient to me.

**Comment by 0dvictor** [comment]
Definitely, and in fact it is exactly what the current version of #67195 is doing.

I’m open to either approach.

**Comment by Mark-Simulacrum** [comment]
Great. Let's close this then and go with that approach, I think it's better -- at least for now.
